File: V2/domain-pipeline-schema/src/nodes/node4.ts
/**
 * Node 4 — Spec Assembly
 *
 * Compiles all prior node outputs into a single artifact.
 * Introduces no new analysis — only compilation and cross-referencing.
 *
 * The Node4Output is the STORED format (JSON). The human-readable markdown
 * is a rendering concern handled by a separate renderer module.
 *
 * Validation layers:
 *   L1 (schema): structural validity
 *   L2 (referential): file index req IDs resolve; all prior IDs preserved
 *   L3 (semantic): metrics match actual counts; every requirement appears
 *                   in file index; every requirement appears in roadmap;
 *                   no data lost from prior nodes
 */

import { z } from "zod";
import { PipelineEnvelope, ReqID } from "../primitives.js";

// ── File Index ──────────────────────────────────────────────────────────────
// Maps source files to the requirements that touch them.
// Built by scanning requirement entries for file references — not guessed.

export const FileIndexEntry = z.object({
  file: z.string().min(1),
  category: z.enum([
    "backend-router",
    "backend-service",
    "backend-agent",
    "frontend-hook",
    "frontend-component",
    "frontend-state",
    "config",
    "schema",
    "shared",
    "other",
  ]),
  requirements: z.array(ReqID).min(1),
});
export type FileIndexEntry = z.infer<typeof FileIndexEntry>;

// ── Changelog ───────────────────────────────────────────────────────────────

export const ChangelogEntry = z.object({
  date: z.string().regex(/^\d{4}-\d{2}-\d{2}$/),
  change: z.string().min(1),
});

// ── Node 4 Output ───────────────────────────────────────────────────────────
// The compiled spec as structured data. Contains references to prior node
// outputs but does NOT duplicate them inline — the prior JSONs remain the
// source of truth. Node 4 adds cross-cutting concerns: file index, changelog,
// computed metrics, and renders to markdown.

export const Node4Output = PipelineEnvelope.extend({
  node: z.literal(4),

  /** Summary of system purpose — one paragraph, not re-analysis. */
  systemPurpose: z.string().min(1),

  /** File index: which files are touched by which requirements. */
  fileIndex: z.array(FileIndexEntry),

  /** Initialised with baseline entry. */
  changelog: z.array(ChangelogEntry).min(1),

  /**
   * Computed metrics. Validated against actual counts from prior nodes.
   * Any mismatch = assembly error.
   */
  metrics: z.object({
    totalGoals: z.number().int(),
    totalAggregates: z.number().int(),
    totalEvents: z.number().int(),
    totalGaps: z.number().int(),
    confirmedObstacles: z.number().int(),
    newObstacles: z.number().int(),
    totalRequirements: z.number().int(),
    must: z.number().int(),
    should: z.number().int(),
    could: z.number().int(),
    phases: z.number().int(),
    filesIndexed: z.number().int(),
  }),

  /**
   * References to the source node outputs.
   * In storage, these could be URIs or object IDs.
   * In local execution, these are file paths.
   */
  sources: z.object({
    node0: z.string().min(1),
    node1: z.string().min(1),
    node2: z.string().min(1),
    node3: z.string().min(1),
  }),
});

export type Node4Output = z.infer<typeof Node4Output>;

File: v3_python/ede/nodes/node4.py
"""
Node 4 -- Spec Assembly

Compiles all prior node outputs into a single artifact.
No new analysis -- only compilation and cross-referencing.

The Node4Output is the STORED format (JSON). Human-readable markdown
is a rendering concern handled separately.

Validation layers:
  L1 (schema): structural validity
  L2 (referential): file index req IDs resolve; all prior IDs preserved
  L3 (semantic): metrics match actual counts; every requirement appears
                  in file index and roadmap
"""

from __future__ import annotations

from enum import StrEnum
from typing import Literal

from pydantic import Field, StringConstraints
from typing import Annotated

from ..primitives import EdeBaseModel, PipelineEnvelope, ReqID


class FileCategory(StrEnum):
    BACKEND_ROUTER = "backend-router"
    BACKEND_SERVICE = "backend-service"
    BACKEND_AGENT = "backend-agent"
    FRONTEND_HOOK = "frontend-hook"
    FRONTEND_COMPONENT = "frontend-component"
    FRONTEND_STATE = "frontend-state"
    CONFIG = "config"
    SCHEMA = "schema"
    SHARED = "shared"
    OTHER = "other"


class FileIndexEntry(EdeBaseModel):
    file: str = Field(min_length=1)
    category: FileCategory
    requirements: list[ReqID] = Field(min_length=1)


class ChangelogEntry(EdeBaseModel):
    date: Annotated[str, StringConstraints(pattern=r"^\d{4}-\d{2}-\d{2}$")]
    change: str = Field(min_length=1)


class Node4Metrics(EdeBaseModel):
    total_goals: int
    total_aggregates: int
    total_events: int
    total_gaps: int
    confirmed_obstacles: int
    new_obstacles: int
    total_requirements: int
    must: int
    should: int
    could: int
    phases: int
    files_indexed: int


class SourceRefs(EdeBaseModel):
    node0: str = Field(min_length=1)
    node1: str = Field(min_length=1)
    node2: str = Field(min_length=1)
    node3: str = Field(min_length=1)


class Node4Output(PipelineEnvelope):
    node: Literal[4]
    system_purpose: str = Field(min_length=1)
    file_index: list[FileIndexEntry]
    changelog: list[ChangelogEntry] = Field(min_length=1)
    metrics: Node4Metrics
    sources: SourceRefs
